server/services/v1/database/secondary_indexer.go

```diff
@@ -0,0 +1,489 @@
+		size, errSize = q.aggregateInfo(ctx, tx, lkeySize, rkeySize)
+	}()
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		lkeyCount := keys.NewKey(q.coll.EncodedName, q.coll.Indexes.SecondaryIndex.Name, InfoSubspace, CountSubSpace, "")
+		rkeyCount := keys.NewKey(q.coll.EncodedName, q.coll.Indexes.SecondaryIndex.Name, InfoSubspace, CountSubSpace, 0xFF)
+		rows, errCount = q.aggregateInfo(ctx, tx, lkeyCount, rkeyCount)
+	}()
+	wg.Wait()
+
+	if errCount != nil {
+		return nil, errCount
+	}
+	if errSize != nil {
+		return nil, errSize
+	}
+
+	return &SecondaryIndexInfo{
+		rows,
+		size,
+	}, nil
+}
+
+func (q *SecondaryIndexer) aggregateInfo(ctx context.Context, tx transaction.Tx, lkey keys.Key, rkey keys.Key) (int64, error) {
+	counter := int64(0)
+	iter, err := tx.AtomicReadRange(ctx, lkey, rkey, false)
+	if err != nil {
+		return 0, err
+	}
+
+	var val kv.FdbBaseKeyValue[int64]
+	for iter.Next(&val) {
+		counter += val.Data
+	}
+
+	if iter.Err() != nil {
+		return 0, iter.Err()
+	}
+
+	return counter, nil
+}
+
+func (q *SecondaryIndexer) Delete(ctx context.Context, tx transaction.Tx, td *internal.TableData, primaryKey []interface{}) error {
+	return q.Update(ctx, tx, nil, td, primaryKey)
+}
+
+func (q *SecondaryIndexer) Index(ctx context.Context, tx transaction.Tx, td *internal.TableData, primaryKey []interface{}) error {
+	return q.Update(ctx, tx, td, nil, primaryKey)
+}
+
+func (q *SecondaryIndexer) Update(ctx context.Context, tx transaction.Tx, newTd *internal.TableData, oldTd *internal.TableData, primaryKey []interface{}) error {
+	updateSet := q.buildAddAndRemoveKVs(newTd, oldTd, primaryKey)
+
+	// Update Row Count
+	err := incAtomicMap(ctx, tx, q.coll, genRowCountKey, updateSet.addCounts)
+	if err != nil {
+		return err
+	}
+	err = decAtomicMap(ctx, tx, q.coll, genRowCountKey, updateSet.removeCounts)
+	if err != nil {
+		return err
+	}
+
+	// Update Row Size
+	err = incAtomicMap(ctx, tx, q.coll, genRowSizeKey, updateSet.addSizes)
+	if err != nil {
+		return err
+	}
+	err = decAtomicMap(ctx, tx, q.coll, genRowSizeKey, updateSet.removeSizes)
+	if err != nil {
+		return err
+	}
+
+	for _, indexKey := range updateSet.removeKeys {
+		err := tx.Delete(ctx, indexKey)
+		if err != nil {
+			return err
+		}
+	}
+
+	for _, indexKey := range updateSet.addKeys {
+		err := tx.Replace(ctx, indexKey, nil, false)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (q *SecondaryIndexer) buildAddAndRemoveKVs(newTableData *internal.TableData, oldTableData *internal.TableData, primaryKey []interface{}) *IndexerUpdateSet {
+	newRows := q.buildTableRows(newTableData)
+	oldRows := q.buildTableRows(oldTableData)
+
+	// Remove any rows in the oldRow set that will be overwritten in the newRows
+	// We do not want to issue an unneeded delete when it will be overwritten with the update
+	rowsToRemove := removeDuplicateRows(newRows, oldRows)
```

Comment:
Your logic of removing will works fine in case of replacement. But in the case of partial updates, it will remove the rows for the fields that are not passed in the new data. For updates, we need to only remove rows that are changed in the new payload i.e. 
- New row should have the field which is also present in the old row. 
- The value in the new row for that field shouldn't match the value in the old row.

Reply:
I've written a few tests here https://github.com/tigrisdata/tigris/blob/cb0c33395786c2b81be8d0bc063337f3a6870706/server/services/v1/database/secondary_indexer_test.go#L90

To test this logic. The basic process is:
1. Build key values for old and new doc
2. Remove keys from the old doc that are exactly the same in the new doc
3. Remove keys from the new doc that are exactly the same as the old doc
4. Remove old keys from the index
4. Add new keys into index

Reply:
So I think you are cleaning up the old fields as well. Can we have a test for this case,
```
old := `{"a": 1, "b": 2, "c": "foo", "d": {"d1": 1, "d2": "foo"}, "e": {"e1": "1", "e2": "foo"}}`
new := `{"a: 2, "d.f1": 2, "e": {"e1": 2}}`
```

Outcome,
```
a-->2
d.f1-->2
d.d2-->foo
e.e1 --> 2
```

- `b, c` are retained from the old doc
- `d.d2` is retained from the old doc
- `e` is completely replaced by the user so only `e.e1`
- `a` is updated.

Reply:
Just one correction, `new` I am referring to the request payload for the Update request from the user.